### ISSUE TITLE
should we try to decode CIOOS-style variable names in NetCDF files?

### ISSUE THREAD

**Issue by dankelley** [dankelley/oce#2235]
(This might apply to other files from CIOOS, too.)

The names are a bit confusing, e.g. in the output shown below, we see that the values for the temperature variables `TEMPS901`, `TEMPP901` and `TEMPPR01` seem to be identical.  But maybe they are not always the same.  So, it would be nice to know which is appropriate for say `as.ctd()` to convert an object with such names into a standard ctd object.

Oh, and a similar thing is going on for `PSALST01` and `PSLTZZ01`: the summary shows identical ranges and means.

I've spent some time looking at the ncdf file, and if your eyes are even a bit tired, it's not easy to spot the differences.

Thankfully, @clayton33 has provided me with a spreadsheet with name translations, so I ought to be able to make progress.

I'll start with temperature.

Here's a checklist, to which I'll add as I get ready to make changes.  Ticking off means only that I've done this in my local code, **not** that I've pushed to github.

- [ ] TEMPS901 -> temperature http://vocab.nerc.ac.uk/collection/P01/current/TEMPS901/ "Unavailable"
- [ ] TEMPP901 -> temperature http://vocab.nerc.ac.uk/collection/P01/current/TEMPP901/ "The degree of hotness at any position within a volume of fresh or salt water expressed against the International Temperature Scale 1990 (ITS-90) calibration standard."
- [ ] TEMPPR01 -> temperature http://vocab.nerc.ac.uk/collection/P01/current/TEMPPR01/ "The degree of hotness of the water column expressed against a standard scale. Includes both IPTS-68 and ITS-90 scales."
- [ ] PRESPR01 -> pressure http://vocab.nerc.ac.uk/collection/P01/current/PRESPR01/ "The force per unit area exerted by the water column on a mobile sensor located in the water column."
- [ ] PSALST01 -> salinity http://vocab.nerc.ac.uk/collection/P01/current/PSALST01/ "This is the preferred term for this definition. Alternative term PSALST02 is included to cover cases where there are two sensors of the same type contributing to the data set and referential integrity considerations prevent a usage of a single code."
 - [ ] PSLTZZ01 -> salinity http://vocab.nerc.ac.uk/collection/P01/current/PSLTZZ01/ "The quantity of dissolved ions (predominantly salt in seawater) expressed on a scale (PSS-78) based on the conductivity ratio of a seawater sample to a standard KCl solution."

<details>
```
R version 4.4.1 (2024-06-14) -- "Race for Your Life"
Copyright (C) 2024 The R Foundation for Statistical Computing
Platform: x86_64-apple-darwin20

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

[Previously saved workspace restored]

> library(oce)
Loading required package: gsw
> # https://cioosatlantic.ca/erddag/files/bio_maritimes_region_ecosystem_survey_ctd/Maritimes%20Region%20Ecosystem%20Survey%20Summer/2023/CTD_CAR2023011_242_497961_UP.ODF.nc
> f <- "~/CTD_CAR2023011_242_497961_UP.ODF.nc"
> options(warn = 1) # so we see warnings as they occur
> d <- read.netcdf(f)
> summary(d)
* Time: 2023-08-14 04:31:09
* Data Overview

                                       Min.                Mean                    Max.                Dim. NAs
    measurement_time                   2023-08-14 05:04:11 2023-08-14 05:13:45.144 2023-08-14 05:26:15 297  0  
    ScanNumber                         3967                5114.283                6614                297  0  
    QCNTR_01                           1                   1                       1                   297  0  
    PRESPR01 [dbar]                    0.5                 74.5                    148.5               297  0  
    QPRES_01                           0                   0                       0                   297  0  
    TEMPS901 [°C, ITS-90]              2.9981              7.232082                19.4415             297  0  
    TEMPP901 [°C, ITS-90]              2.9981              7.232082                19.4415             297  0  
    TEMPPR01 [°C, ITS-90]              2.9981              7.232082                19.4415             297  0  
    QTEMP_01                           1                   1                       1                   297  0  
    CNDCST01 [S/m]                     2.9469              3.346433                4.225588            297  0  
    QCNDC_01                           1                   1                       1                   297  0  
    OXYOCPVL01 [V]                     1.8417              2.282159                2.98                297  0  
    QOXYV_01                           1                   1                       1                   297  0  
    IRRDUV01 [μEinstein/s/kg]          0                   0.01430488              0.63693             297  0  
    QPSAR_01                           1                   1                       1                   297  0  
    CPHLPR01 [mg/m³]                   0.5206              2.057365                9.3864              297  1  
    QCPHLPR01                          1                   1.026936                9                   297  0  
    AHSFZZ01 [m]                       2.78                57.74862                102.47              297  0  
    QALTB_01                           1                   1.040404                4                   297  0  
    preciseLat [°N]                    44.27               44.27085                44.2719             297  0  
    QLATD_01                           0                   0                       0                   297  0  
    preciseLon [°E]                    -63.319             -63.3185                -63.31844           297  0  
    QLOND_01                           0                   0                       0                   297  0  
    PSALST01 [PSS-78]                  30.773              32.72341                34.1996             297  0  
    PSLTZZ01 [PSS-78]                  30.773              32.72341                34.1996             297  0  
    QPSAL_01                           1                   1                       1                   297  0  
    POTMCV01 [°C, ITS-90]              2.9953              7.225871                19.4404             297  0  
    QPOTM_01                           1                   1                       1                   297  0  
    SIGTEQ01 [kg/m³]                   21.71               25.48022                26.5774             297  0  
    QSIGP_01                           1                   1                       1                   297  0  
    DOXYZZ01 [ml/l]                    4.2994              5.936503                7.524               297  0  
    QDOXY_01                           1                   1                       1                   297  0  
    RecPerBin                          2                   5.552189                18                  297  0  
    QCNTR_02                           1                   1                       1                   297  0  
    QCFF_01                            0                   0.1616162               12                  297  0  
    time                               2023-08-14 04:31:09 2023-08-14 04:31:09     2023-08-14 04:31:09 1    0  
    latitude [°N]                      44.269              44.2685                 44.2685             1    0  
    longitude [°E]                     -63.319             -63.3195                -63.3195            1    0  
    TemperatureSensor                  NA                  NA                      NA                  1    0  
    ConductivitySensor                 NA                  NA                      NA                  1    0  
    PressureSensor                     NA                  NA                      NA                  1    0  
    OxygenSensor                       NA                  NA                      NA                  1    0  
    PAR_BiosphericalLicorChelseaSensor NA                  NA                      NA                  1    0  
    FluoroWetlabECO_AFL_FL_Sensor      NA                  NA                      NA                  1    0  
    AltimeterSensor                    NA                  NA                      NA                  1    0  

* Processing Log

    - 2024-08-22 19:16:59.015 UTC: `Create oce object`
    - 2024-08-22 19:16:59.098 UTC: `read.netcdf("~/CTD_CAR2023011_242_497961_UP.ODF.nc")`
> 
```

**Comment by dankelley** [comment]
TEMPP901: see http://vocab.nerc.ac.uk/collection/P01/current/TEMPP901/

**Comment by richardsc** [comment]
Not advocating for this per se, but I feel it's worth at least stating and discussing, but what about handling specific ncdf format oceanographic files in _add-on_ packages to `oce`, instead of hard-baking all the fiddly naming stuff into it? Certainly makes it easier from a build/test perspective, and also more likely to have others (e.g. CIOOS employees?) contribute.

TL;DR: Should we make a "cioos" package that converts known ncdf formats to oce objects?

**Comment by dankelley** [comment]
Good to discuss in person sometime, regarding the intricacies of e.g. setting up the original-name system for summaries and `[[`.

What I've done for `read.netcdf()` is to add an extra parameter called `renamer`, which is a function.  Users can make their own, or just extend the existing by writing a function that extends those cases.  FYI at the moment I have as follows.  I don't think it's quite right, though, because some names are not documented on the NERC webpages.

```R
#' Determine oce Variable Names from an NERC/BODC Names
#'
#' Find variable names from Reference 1.  This handles only a very
#' tiny subset of the huge number of names in the NERC/BODC vocabulary
#' (e.g. several dozens of salinity variants). To see the
#' handled names, type `bodcName2oceName` in an R session, which
#' will reveal the code.
#'
#' Note to developers: it is relatively easy to add new names, because
#' reasonably thorough documentation is available online, at a website that
#' includes the variable name. For example,
#' <http://vocab.nerc.ac.uk/collection/P01/current/PRESPR01/> provides
#' information on the `PRESPR01` variable (a form of pressure). Since the
#' NetCDF file contains information about the units, you need not incorporate
#' them into the variable name. For example, in the code you can see that
#' several variables get mapped to the oce name `temperature`.
#'
#' @param bodcNames character vector that specifies variable names
#' that use the NERC/BODC convention.
#'
#' @return
#' `bodcNames2oceNames` returns a vector of the same length as its
#' first argument, with translations to oce names, as appropriate.
#' Note that the usual oce convention for handling duplicates is
#' used, with the first name that maps to temperature being set
#' to `temperature`, the next to `temperature2`, etc.
#'
#' @examples
#' bodcNames2oceNames(c("PSALST01", "TEMPP901", "PRESPR01"))
#'
#' @references
#' 1. The NERC Environmental Data Server.
#' <http://vocab.nerc.ac.uk/collection/P01/current/>
#'
#' @family functions that convert variable names to the oce convention
#'
#' @author Dan Kelley
bodcNames2oceNames <- function(bodcNames) {
    rval <- bodcNames
    for (i in seq_along(bodcNames)) {
        if (bodcNames[i] %in% c("PRESPR01")) {
            rval[i] <- "pressure"
        } else if (bodcNames[i] %in% c("PSALST01", "PSLTZZ01")) {
            rval[i] <- "salinity"
        } else if (bodcNames[i] %in% c(
            "TEMPP681", "TEMPS601", "TEMPR601",
            "TEMPPR01"
        )) {
            rval[i] <- "temperature" # "degree C", "IPTS-68"
        } else if (bodcNames[i] %in% c("TEMPP901", "TEMPS901", "TEMPS902", "TEMPR901")) {
            rval[i] <- "temperature" # "degree C", "ITS-90"
        } else if (bodcNames[i] %in% c("DOXYZZ01")) {
            rval[i] <- "oxygen"
        } else if (bodcNames[i] %in% c("OXYOCPVL01")) {
            rval[i] <- "oxygenVoltage"
        } else if (bodcNames[i] %in% c("CPHLPR0", "CPHLPR01")) {
            rval[i] <- "chlorophyll-a"
        } else if (bodcNames[i] %in% c("ScanNumber")) {
            rval[i] <- "scan"
        } else if (bodcNames[i] %in% c("CNDCST01")) {
            rval[i] <- "conductivity"
        } else if (bodcNames[i] %in% c("IRRDUV01")) {
            rval[i] <- "downwellingIrradiance"
        } else if (bodcNames[i] %in% c("measurement_time")) {
            rval[i] <- "measurementTime"
        } else if (bodcNames[i] %in% c("AHSFZZ01")) {
            rval[i] <- "heightAboveBottom"
        } else if (bodcNames[i] %in% c("POTMCV01")) {
            rval[i] <- "theta" # unesco
        } else if (bodcNames[i] %in% c("SIGTEQ01")) {
            rval[i] <- "sigmaTheta" # unesco
        }
    }
    unduplicateNames(rval)
}
```

**Comment by dankelley** [comment]
This is done in commit 4246e4b8bcb22c516d8e5e7f3ec1ea743b0252f3 of branch "develop", with as follows on a test file.
```R

R version 4.4.1 (2024-06-14) -- "Race for Your Life"
Copyright (C) 2024 The R Foundation for Statistical Computing
Platform: x86_64-apple-darwin20

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> library(oce)
> # https://cioosatlantic.ca/erddag/files/bio_maritimes_region_ecosystem_survey_ctd/Maritimes%20Region%20Ecosystem%20Survey%20Summer/2023/CTD_CAR2023011_242_497961_UP.ODF.nc
> f <- "~/CTD_CAR2023011_242_497961_UP.ODF.nc"
> d <- read.netcdf(f)
> summary(d)
* Time: 2023-08-14 04:31:09
* Data Overview

                                       Min.                Mean                    Max.                Dim. NAs OriginalName                        
    measurement_time                   2023-08-14 05:04:11 2023-08-14 05:13:45.144 2023-08-14 05:26:15 297  0   "measurement_time"                  
    ScanNumber                         3967                5114.283                6614                297  0   "ScanNumber"                        
    QCNTR_01                           1                   1                       1                   297  0   "QCNTR_01"                          
    PRESPR01 [dbar]                    0.5                 74.5                    148.5               297  0   "PRESPR01"                          
    QPRES_01                           0                   0                       0                   297  0   "QPRES_01"                          
    TEMPS901 [°C, ITS-90]              2.9981              7.232082                19.4415             297  0   "TEMPS901"                          
    TEMPP901 [°C, ITS-90]              2.9981              7.232082                19.4415             297  0   "TEMPP901"                          
    TEMPPR01 [°C, ITS-90]              2.9981              7.232082                19.4415             297  0   "TEMPPR01"                          
    QTEMP_01                           1                   1                       1                   297  0   "QTEMP_01"                          
    CNDCST01 [S/m]                     2.9469              3.346433                4.225588            297  0   "CNDCST01"                          
    QCNDC_01                           1                   1                       1                   297  0   "QCNDC_01"                          
    OXYOCPVL01 [V]                     1.8417              2.282159                2.98                297  0   "OXYOCPVL01"                        
    QOXYV_01                           1                   1                       1                   297  0   "QOXYV_01"                          
    IRRDUV01 [μEinstein/s/kg]          0                   0.01430488              0.63693             297  0   "IRRDUV01"                          
    QPSAR_01                           1                   1                       1                   297  0   "QPSAR_01"                          
    CPHLPR01 [mg/m³]                   0.5206              2.057365                9.3864              297  1   "CPHLPR01"                          
    QCPHLPR01                          1                   1.026936                9                   297  0   "QCPHLPR01"                         
    AHSFZZ01 [m]                       2.78                57.74862                102.47              297  0   "AHSFZZ01"                          
    QALTB_01                           1                   1.040404                4                   297  0   "QALTB_01"                          
    preciseLat [°N]                    44.27               44.27085                44.2719             297  0   "preciseLat"                        
    QLATD_01                           0                   0                       0                   297  0   "QLATD_01"                          
    preciseLon [°E]                    -63.319             -63.3185                -63.31844           297  0   "preciseLon"                        
    QLOND_01                           0                   0                       0                   297  0   "QLOND_01"                          
    PSALST01 [PSS-78]                  30.773              32.72341                34.1996             297  0   "PSALST01"                          
    PSLTZZ01 [PSS-78]                  30.773              32.72341                34.1996             297  0   "PSLTZZ01"                          
    QPSAL_01                           1                   1                       1                   297  0   "QPSAL_01"                          
    POTMCV01 [°C, ITS-90]              2.9953              7.225871                19.4404             297  0   "POTMCV01"                          
    QPOTM_01                           1                   1                       1                   297  0   "QPOTM_01"                          
    SIGTEQ01 [kg/m³]                   21.71               25.48022                26.5774             297  0   "SIGTEQ01"                          
    QSIGP_01                           1                   1                       1                   297  0   "QSIGP_01"                          
    DOXYZZ01 [ml/l]                    4.2994              5.936503                7.524               297  0   "DOXYZZ01"                          
    QDOXY_01                           1                   1                       1                   297  0   "QDOXY_01"                          
    RecPerBin                          2                   5.552189                18                  297  0   "RecPerBin"                         
    QCNTR_02                           1                   1                       1                   297  0   "QCNTR_02"                          
    QCFF_01                            0                   0.1616162               12                  297  0   "QCFF_01"                           
    time                               2023-08-14 04:31:09 2023-08-14 04:31:09     2023-08-14 04:31:09 1    0   "time"                              
    latitude [°N]                      44.269              44.2685                 44.2685             1    0   "latitude"                          
    longitude [°E]                     -63.319             -63.3195                -63.3195            1    0   "longitude"                         
    TemperatureSensor                  NA                  NA                      NA                  1    0   "TemperatureSensor"                 
    ConductivitySensor                 NA                  NA                      NA                  1    0   "ConductivitySensor"                
    PressureSensor                     NA                  NA                      NA                  1    0   "PressureSensor"                    
    OxygenSensor                       NA                  NA                      NA                  1    0   "OxygenSensor"                      
    PAR_BiosphericalLicorChelseaSensor NA                  NA                      NA                  1    0   "PAR_BiosphericalLicorChelseaSensor"
    FluoroWetlabECO_AFL_FL_Sensor      NA                  NA                      NA                  1    0   "FluoroWetlabECO_AFL_FL_Sensor"     
    AltimeterSensor                    NA                  NA                      NA                  1    0   "AltimeterSensor"                   

* Processing Log

    - 2024-08-23 17:20:20.974 UTC: `Create oce object`
    - 2024-08-23 17:20:21.066 UTC: `read.netcdf("~/CTD_CAR2023011_242_497961_UP.ODF.nc")`
> 
```
